### PR TITLE
remove user specification via request parameter for TSB

### PR DESCRIPTION
**Pull request by gabemontero**
@openshift/devex ptal

@pmorie @staebler fyi

### CODE REVIEW

**Comment by gabemontero** [comment]
ah ... was to mention @spadgett but the robot beat me to it :-)

**Comment by gabemontero** [comment]
Hmm ... import verifier is flagging an import I'm not immediately seeing:

```
2017/09/19 15:55:31 Inspecting imports under github.com/openshift/origin/pkg/templateservicebroker...
2017/09/19 15:55:32 -- validating imports for 5 packages in the tree
2017/09/19 15:55:32 -- found unused package imports
2017/09/19 15:55:32 	github.com/openshift/origin/pkg/user/apis/user/validation
```

investigating ...

**Comment by gabemontero** [comment]
the user validate import we now no longer require needed to be removed from hack/import-restrictions.json ... change pushed

**Comment by gabemontero** [comment]
integration tests saw an apparent reincarnation of https://github.com/openshift/origin/issues/14020

**Comment by gabemontero** [comment]
@spadgett and I gave this a go and realized we didn't enable the orig id hdr feature via the new controller manager mechanism.

We didn't see a way to actually do that yet via a `oc cluster up --version=latest --service-catalog` path

@pmorie @staebler - are we correct in that there is not a way to do that yet?  Or if there is, please provide us the details?

thanks

**Comment by staebler** [comment]
@gabemontero @spadgett You are correct that there is not currently a way to enable the feature. That was missed but has been checked in. There will be a new release of service catalog tonight that will include the flag to enabled feature gates in controller manager. When it is ready, you will be able to enable the feature with the argument `--feature-gates OriginatingIdentity=true`.

**Comment by gabemontero** [comment]
I believe I sorted out the integration test issue ... `may not contain \":\""` is the key.  Most likely we are coming in as either`system:authenticated` or `system:anonymous` and the username validation has failed.

We discussed this some before during the original pull ... I believe the net was long term, those two username's are valid.  Bottom line, I believe are validation should stop prior to sorting out the specifics of the username, and let the lower level auth stuff manage it.

**Comment by gabemontero** [comment]
On Tue, Sep 19, 2017 at 4:14 PM, Matthew Staebler <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> @spadgett
> <https://github.com/spadgett> You are correct that there is not currently
> a way to enable the feature. That was missed but has been checked in. There
> will be a new release of service catalog tonight that will include the flag
> to enabled feature gates in controller manager. When it is ready, you will
> be able to enable the feature with the argument --feature-gates
> OriginatingIdentity=true.
>

Thanks for the confirmation and update @staebler.

**Comment by gabemontero** [comment]
ok integration tests are now passing, but somewhere along the way since first starting with this some of the template related extended tests are now failing; need to circle back and clean that back up;

File: pkg/template/registry/templateinstance/strategy.go [hunk 1]
```diff
@@ -52,14 +52,9 @@ func (templateInstanceStrategy) Canonicalize(obj runtime.Object) {
 func (templateInstanceStrategy) PrepareForCreate(ctx apirequest.Context, obj runtime.Object) {
 	templateInstance := obj.(*templateapi.TemplateInstance)
 
-	//TODO - when https://github.com/kubernetes-incubator/service-catalog/pull/939 sufficiently progresses, remove the if nil check and always
-	// Clear any user-specified info, then seed it from the context
-	if templateInstance.Spec.Requester == nil {
```
Comment (jim-minter) [inline pkg/template/registry/templateinstance/strategy.go]: Could you remind me on what basis this is being removed?  Does it affect the TSB setting the templateinstance requester on behalf of the TSB caller?  If so, a test should be tripped?

Comment (gabemontero) [reply pkg/template/registry/templateinstance/strategy.go]: Actually @jim-minter perhaps my TODO comment here was erroneous, given the test failure noted at https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/16434/test_pull_request_origin_extended_conformance_gce/8233/

Or maybe an associated change elsewhere is needed.

I've started investigating what I currently have here, and cycling through the original PR.  Will report back when I have discoveries of note.

**Comment by gabemontero** [comment]
ok ... all happy tests .... the strategy.go change was the last bit

will ask for a quick re-review in code

File: pkg/template/registry/templateinstance/strategy.go [hunk 2]
```diff
@@ -52,8 +52,9 @@ func (templateInstanceStrategy) Canonicalize(obj runtime.Object) {
 func (templateInstanceStrategy) PrepareForCreate(ctx apirequest.Context, obj runtime.Object) {
 	templateInstance := obj.(*templateapi.TemplateInstance)
 
-	//TODO - when https://github.com/kubernetes-incubator/service-catalog/pull/939 sufficiently progresses, remove the if nil check and always
-	// Clear any user-specified info, then seed it from the context
+	// if request not set, pull from context; note: the requester can be set via the service catalog
+	// propagating the user information via the openservicebroker origination api header on
+	// calls to the TSB endpoints (i.e. the Provision call)
 	if templateInstance.Spec.Requester == nil {
```
Comment (gabemontero) [inline pkg/template/registry/templateinstance/strategy.go]: @liggitt @pmorie - if you all could, please re-review / take a quick peek at `PrepareForCreate` and see if you agree.  I tried to capture in the comment above what we think is different from "typical" `PrepareForCreate`.  Namely:
- the entry points like `Provision` into the TSB will provide the user.Info (as the service catalog propagates it in the orig identity header) and we set the user onto the `templateInstance.Spec.Requester` there
- so if it is set, we don't nil out
- if it is not set, we pull from the context

Ultimately, we test this behavior in various places, most notably our impersonation tests:  https://github.com/openshift/origin/blob/master/test/extended/templates/templateinstance_impersonation.go#L288

I believe prior agreement was obtained on all this, but after some discussions with @jim-minter it seemed prudent to double-check one more time.

thanks

@bparees fyi

**Comment by gabemontero** [comment]
pending the one confirmation request to @liggitt @pmorie I'll remove the do-not-merge label once I'm able to provision a template from the console with the `--feature-gates OriginatingIdentity=true` option added to the service catalog template.

**Comment by liggitt** [comment]
> if it is set, we don't nil out

This means we can never give access to this API to a non-superuser, right?

**Comment by jim-minter** [comment]
> This means we can never give access to this API to a non-superuser, right?

Ability to set the username to anything that isn't your username should be being gated in `func (s *templateInstanceStrategy) validateImpersonation`.

**Comment by spadgett** [comment]
@gabemontero #16457 should enable the originating identity header

**Comment by gabemontero** [comment]
thanks @spadgett / yep I had been monitoring that PR, as well as https://github.com/openshift/origin/pull/16440 before it

regardless, I was interested in trying it manually anyway :-)

I just rebased, and pulled down the following service catalog image:

```
gmontero ~/go/src/github.com/openshift/origin  (finish-tsb-group-auth)$ docker images | grep catalog
docker.io/openshift/origin-service-catalog         latest              4e745d14048f        About an hour ago   361.3 MB
gmontero ~/go/src/github.com/openshift/origin  (finish-tsb-group-auth)$ 
```

And manually updated the template ... here is the relevant snippet:

```
        - command: 
          - apiserver
          args:
          - --admission-control
          - KubernetesNamespaceLifecycle,DefaultServicePlan,ServiceInstanceCredentialsLifecycle,ServicePlanChangeValidator
          - --storage-type
          - etcd
          - --secure-port
          - "6443"
          - --etcd-servers
          - http://localhost:2379
          - -v
          - "10"
          - --cors-allowed-origins
          - ${CORS_ALLOWED_ORIGIN}
          - --feature-gates OriginatingIdentity=true
```

But the service catalog reports this error:

```
2017-09-21T14:22:07.442603114Z E0921 14:22:07.442113       1 apiserver.go:50] server exited unexpectedly (unknown flag: --feature-gates OriginatingIdentity)
```

@staebler @jboyd01 @pmorie - any insight if I did something wrong?  Or are there additional changes I need to wait on?

thanks

**Comment by staebler** [comment]
@gabemontero My understanding with the helm chart is that you need `--feature-gates` on one line and `OriginatingIdentity=true` on the next line.

```
        - command: 
          - apiserver
          args:
          - --admission-control
          - KubernetesNamespaceLifecycle,DefaultServicePlan,ServiceInstanceCredentialsLifecycle,ServicePlanChangeValidator
          - --storage-type
          - etcd
          - --secure-port
          - "6443"
          - --etcd-servers
          - http://localhost:2379
          - -v
          - "10"
          - --cors-allowed-origins
          - ${CORS_ALLOWED_ORIGIN}
          - --feature-gates
          - OriginatingIdentity=true
```

**Comment by spadgett** [comment]
@gabemontero I don't think that image has the changes in 0.0.21. You need to build a new service catalog image from #16457

**Comment by gabemontero** [comment]
Thanks for the quick resposnes @staebler and @spadgett 

And yep,
1) splitting `--feature-gates` and `OriginatingIdentity=true` at least allowed the svc catalog to start up
2) and without https://github.com/openshift/origin/pull/16457 (which I believe makes sense to me now after perusing it more closely), we get this when attempting to provision a template from the console:

```
Error provisioning ServiceInstance "myproject/mariadb-persistent-rldw0" of ServiceClass "mariadb-persistent" at ServiceBroker "template-service-broker": Status: 400; ErrorMessage: <nil>; Description: couldn't parse X-Broker-API-Originating-Identity header; ResponseError: <nil>
```

I'm going to stop here for now (move on to other to-do's) and wait for https://github.com/openshift/origin/pull/16457 to progress vs. checking out https://github.com/openshift/origin/pull/16457 and trying to build the service catalog image.

**Comment by staebler** [comment]
@gabemontero I think the next step that you are missing is that the OriginatingIdentity feature needs to be enabled for service-catalog apiserver as well as controller-manager. Re-reading my previous comments, I see where it can be interpreted that it is only needed for controller-manager. To clarify my previous comment, the bug in the v0.0.20 release of service-catalog was that there was no way to enable the OriginatingIdentity feature in the controller-manager. It was possible to enable it in the apiserver.

**Comment by gabemontero** [comment]
OK, I was able to do some basic provision / bind flows rebased on https://github.com/openshift/origin/commit/b3912c0b3adb5e5429bf2f2ece5cfa83ffe43b88

@bparees - if still inclined please re-add you good to merge tag from https://github.com/openshift/origin/pull/16434#issuecomment-330624161

**Comment by bparees** [comment]
@gabemontero has a service catalog image that provides originating identity been published to dockerhub?  we can't merge this until that has happened right?

**Comment by gabemontero** [comment]
On Fri, Sep 22, 2017 at 6:15 PM Ben Parees <notifications@github.com> wrote:

> @gabemontero <https://github.com/gabemontero> has a service catalog image
> that provides originating identity been published to dockerhub?
>
Yes it has and that is what I tested with.  I monitored the
push-origin-images job and pulled the image once the job with the
appropriate commit finished.

> we can't merge this until that has happened right?
>

Correct but per above we are good there

**Comment by crossonrose** [comment]
PWNED!
________________________________

**Pull request merged**